Implement Software Keyboard GTK frontend

mageven opened the pull request:
Implements a GTK Input Dialog box to enter text whenever an app requests Software Keyboard.
No functional changes to SwKbd applet. But, there is proper separation from GUI components. 

This code was written about a month ago for testing purposes (waiting for the eventual swkbd overhaul).

Resolves #1304 

![swkbd-gui-new](https://user-images.githubusercontent.com/62494521/88790560-fb158a00-d1b5-11ea-94b8-493682236105.PNG)
![swkbd-gui-2](https://user-images.githubusercontent.com/62494521/88538517-20c35780-d02d-11ea-911c-729cc2568c03.PNG)

EDIT: Added a few images
EDIT2: Cleaned up description and added new image.


Thanks to @jgoyvaerts for reminding me that this was in my pile (and sorry).

zombie167 commented:
Thank you very much!
Now the game can enter the name correctly

![sshot-1](https://user-images.githubusercontent.com/22722505/88594965-bf36d380-d094-11ea-98cd-d8cce529f3c3.jpg)

ReaperOfSouls1909 commented:
We love you !!! Will test soon

Doctorwho1909 commented:
![image](https://user-images.githubusercontent.com/64975148/88600418-9686ed80-d044-11ea-90d9-e204bb90bcb1.png)
functions well anyway to read how many letters the game can use? and apply it to the pop up so we know the limits

jduncanator commented:
I like the separation here, but I don't really think the ConfigurationState is the correct place to store the callback.

gdkchan commented on Ryujinx.Common/Configuration/ConfigurationState.cs:
```diff
@@ -336,6 +336,11 @@ public GraphicsSection()
+        /// <summary>
+        /// Callback method that SwKbd applet uses to open an Input Dialog box
+        /// </summary>
+        public Func<SoftwareKeyboardAppletArgs, string> SwKbdHandler { get; set; }
```
This shouldn't be part of configuration imo.

gdkchan commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -62,12 +63,18 @@ public ResultCode GetResult()
+                    + $"Must be between {_keyboardConfig.StringLengthMin} and {_keyboardConfig.StringLengthMax} characters long.\nInvalid Characters: {_keyboardConfig.InvalidCharFlag}",
+                GuideText = _keyboardConfig.GuideText,
+                SubmitText = (!string.IsNullOrWhiteSpace(_keyboardConfig.SubmitText) ? _keyboardConfig.SubmitText : "OK"),
+                InitialText = "" // TODO: add this after transfer memory works
```
What's wrong with transfer memory?

mageven replied:
Made it work with a couple of changes in Applet code. Please check if it is enough.

gdkchan commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -62,12 +63,18 @@ public ResultCode GetResult()
-            }
+                HeaderText = _keyboardConfig.HeaderText,
+                SubtitleText = (!string.IsNullOrWhiteSpace(_keyboardConfig.SubtitleText) ?  $"{_keyboardConfig.SubtitleText}\n" : "")
+                    + $"Must be between {_keyboardConfig.StringLengthMin} and {_keyboardConfig.StringLengthMax} characters long.\nInvalid Characters: {_keyboardConfig.InvalidCharFlag}",
```
IMO this should be limited to only pass along the "SubtitleText" that is sent by the game. The GUI can decide if it's necessary to display more information, like number of character or accepted characters. For example, we could implement another GUI where we instead limit the number of characters on the text box, and also do all validation there. The Min/Max length and character filters could be extra fields on the struct.

mageven commented:
Implemented many of @jduncanator's ideas from our discussions in discord dev channel. 
Review comments were also addressed in that discussion.

`ILibraryAppletCreator::CreateTransferMemoryStorage` now copies transfer memory. Initial Text seems to be working.

Added string length validation. Also, moved length info to another label. 
Removed Invalid characters info as games don't seem to mind them.
New window looks like this (Input was filled by the game)-
![swkbd-gui-new](https://user-images.githubusercontent.com/62494521/88790560-fb158a00-d1b5-11ea-94b8-493682236105.PNG)

jduncanator commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -67,6 +67,17 @@ public ResultCode GetResult()
+
+            // Initial Text is encoded as a UTF-16 string in the work buffer (passed as transfer memory)
+            // InitialStringOffset points to the memory offset and InitialStringLength is the number of UTF-16 characters
+            string initialText = Encoding.Unicode.GetString(_transferMemory, _keyboardConfig.InitialStringOffset, 2 * _keyboardConfig.InitialStringLength);
```
I'm fairly certain this is passed as UTF-8 when the UseUTF8 flag is passed in the keyboard config, so you should use `_encoding.GetString` instead.

jduncanator replied:
Nevermind, on further investigation, nnsdk converts from UTF-8 to UTF-16 internally when `SetInitialTextUtf8` is used. The initial text (and all the other swkbd configuration strings) are _always_ UTF-16. UseUtf8 only adjusts the output encoding.

gdkchan commented on Ryujinx/Ui/InputDialog.cs:
```diff
@@ -0,0 +1,67 @@
+                _stringrange = value;
+                OnInputChanged(this, EventArgs.Empty);
+
+                if(_stringrange.Max < 0)
```
```suggestion
                if (_stringrange.Max < 0)
```

gdkchan commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardUiArgs.cs:
```diff
@@ -0,0 +1,12 @@
+        public string InitialText;
+        public string GuideText;
+        public string SubmitText;
+        public (int Min, int Max) AllowedStringSize;
```
Imo it would be better if this was just 2 separate fields instead of using a tuple.

gdkchan commented on Ryujinx/Ui/InputDialog.cs:
```diff
@@ -0,0 +1,67 @@
+
+        public void OnInputChanged(object sender, EventArgs e)
+        {
+            if(_stringrange.Max < 0)
```
```suggestion
            if (_stringrange.Max < 0)
```

gdkchan commented on Ryujinx.HLE/HOS/Services/Am/AppletAE/AllSystemAppletProxiesService/SystemAppletProxy/ILibraryAppletCreator.cs:
```diff
@@ -36,10 +37,21 @@ public ResultCode CreateTransferMemoryStorage(ServiceCtx context)
+            KTransferMemory tm = context.Process.HandleTable.GetObject<KTransferMemory>(khandle);
 
-            MakeObject(context, new IStorage(new byte[size]));
+            if(tm == null)
```
```suggestion
            if (tm == null)
```

gdkchan commented on Ryujinx.HLE/HOS/Services/Am/AppletAE/AllSystemAppletProxiesService/SystemAppletProxy/ILibraryAppletCreator.cs:
```diff
@@ -36,10 +37,21 @@ public ResultCode CreateTransferMemoryStorage(ServiceCtx context)
         {
             bool unknown = context.RequestData.ReadBoolean();
             long size    = context.RequestData.ReadInt64();
+            int khandle  = context.Request.HandleDesc.ToCopy[0];
```
I would just call it `handle`.

gdkchan commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -38,11 +44,15 @@ internal class SoftwareKeyboardApplet : IApplet
+            var keyboardConfig = _normalSession.Pop();
             _keyboardConfig = ReadStruct<SoftwareKeyboardConfig>(keyboardConfig);
 
+            if(!_normalSession.TryPop(out _transferMemory))
```
```suggestion
            if (!_normalSession.TryPop(out _transferMemory))
```

jduncanator commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -62,20 +72,35 @@ public ResultCode GetResult()
+            };
+
+            // Call the configured GUI handler to get user's input
+            _textValue = _device.UiHandler.DisplayInputDialog(args);
```
This should check `UiHandler` is not null, and leave the default `_textValue` if it is.

mageven replied:
I wrapped the call in a generic try catch since it's an external call and can throw in various ways.

gdkchan replied:
If that's the case, I think it should try-catch on the GUI. Ideally, for those public APIs, the exceptions that it may throw should be well documented, otherwise it would be difficult to use it reliably when each UI could be throwing all sorts of exception when the callback is called.

It could be documented to just return null when a error happens, since it's supposed to return a string.

gdkchan commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -76,6 +88,28 @@ private void Execute()
+            {
+                _textValue = _device.UiHandler.DisplayInputDialog(args);
+            }
+            catch(Exception e)
```
```suggestion
            catch (Exception e)
```
Sorry for nitpicking.

mageven commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -76,6 +97,22 @@ private void Execute()
+            };
+
+            // Call the configured GUI handler to get user's input
+            _okPressed = _device.UiHandler.DisplayInputDialog(args, out _textValue);
```
> jduncanator: This should check UiHandler is not null, and leave the default _textValue if it is.

This comment is relevant again since I moved error handling out of applet code.
Not doing a null-check was intentional, because I feel adding a DummyHandler in the future is better than silently ignoring it here (at least until the code-base is stable)

jduncanator replied:
Given that `UiHandler` _can_ be null, for whatever reason, it's appropriate to do an null check. You can just simply wrap it in an `if` statement and leave the default `_textValue` if it's null.

```
if(_device.UiHandler != null)
{
    // ...
}
```

gdkchan commented on Ryujinx.HLE/HOS/Services/Am/AppletAE/AllSystemAppletProxiesService/SystemAppletProxy/ILibraryAppletCreator.cs:
```diff
@@ -36,10 +37,21 @@ public ResultCode CreateTransferMemoryStorage(ServiceCtx context)
+            {
+                Logger.PrintWarning(LogClass.ServiceAm, $"Invalid TransferMemory Handle: {handle:X}");
+
+                return ResultCode.Success; // TODO: Find correct error code
```
If the application tried to pass a invalid handle through IPC, the kernel would return a `InvalidHandle` error on the `SendSyncRequest` syscall (from what I recall). I don't think there's any way to do something like that currently.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -191,6 +193,8 @@ private MainWindow(Builder builder) : base(builder.GetObject("_mainWin").Handle)
 
             _statusBar.Hide();
+
+            _uiHandler ??= new GtkHostUiHandler(this);
```
Why the `??=` here? Isn't `_uiHandler` always going to be null at this point?

mageven replied:
Nice catch. Fixed.

AcK77 commented on Ryujinx.HLE/HOS/Applets/SoftwareKeyboard/SoftwareKeyboardApplet.cs:
```diff
@@ -38,10 +45,22 @@ internal class SoftwareKeyboardApplet : IApplet
             _interactiveSession.DataAvailable += OnInteractiveData;
 
             var launchParams   = _normalSession.Pop();
+
```
Remove extra empty line

AcK77 commented on Ryujinx.HLE/HOS/Services/Am/AppletAE/AllSystemAppletProxiesService/SystemAppletProxy/ILibraryAppletCreator.cs:
```diff
@@ -36,10 +37,21 @@ public ResultCode CreateTransferMemoryStorage(ServiceCtx context)
+            int  handle  = context.Request.HandleDesc.ToCopy[0];
 
-            // NOTE: We don't support TransferMemory for now.
+            KTransferMemory xfm = context.Process.HandleTable.GetObject<KTransferMemory>(handle);
```
I’d use a bit more explicit name here instead of "xfm"

AcK77 commented on Ryujinx/Ui/GtkHostUiHandler.cs:
```diff
@@ -0,0 +1,68 @@
+                    {
+                        Title = "Software Keyboard",
+                        Text = args.HeaderText,
+                        SecondaryText = args.SubtitleText,
```
```suggestion
                        SecondaryText = args.SubtitleText
```

AcK77 commented on Ryujinx/Ui/GtkHostUiHandler.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public bool DisplayInputDialog(SoftwareKeyboardUiArgs args, out string userText)
+        {
+            ManualResetEvent mre = new ManualResetEvent(false);
```
Maybe we should call it « inputTextEvent » or something like that?

AcK77 commented on Ryujinx/Ui/GtkHostUiHandler.cs:
```diff
@@ -0,0 +1,68 @@
+            mre.WaitOne();
+
+            userText = error ? null : inputText;
+            return error || okPressed;
```
Add a new empty line before

AcK77 commented on Ryujinx/Ui/InputDialog.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    public class InputDialog : MessageDialog
+    {
+        private int _inputmin, _inputmax;
```
```suggestion
        private int _inputMin, _inputMax;
```

Kryokina commented:
![SDV_Screenshot](https://user-images.githubusercontent.com/36516893/175104151-a456f8b3-f197-4f04-888d-c48f2dfb7feb.png)
How do you get the Software Keyboard pop-up to go away after you have typed what you want?  Nothing I have tried has worked (pressing OK, Cancel, X, etc.).
